vlasovsolver/vlasovmover.cpp

```diff
@@ -243,6 +240,77 @@ void calculateSpatialTranslation(
    // bailout(true, "", __FILE__, __LINE__);
 }
 
+/** Propagates the distribution function in spatial space.
+    Now does extra calculations locally without interim MPI communication.
+
+    Based on SLICE-3D algorithm: Zerroukat, M., and T. Allen. "A
+    three‐dimensional monotone and conservative semi‐Lagrangian scheme
+    (SLICE‐3D) for transport problems." Quarterly Journal of the Royal
+    Meteorological Society 138.667 (2012): 1640-1651.
+
+ */
+void calculateSpatialLocalTranslation(
+   dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>& mpiGrid,
+   const vector<CellID>& local_propagated_cells, // Used for loadbalancing, not selecting active cells
+   const vector<CellID>& remoteTargetCellsx, // Not necessary, passed for compatibility
+   const vector<CellID>& remoteTargetCellsy, // Not necessary, passed for compatibility
+   const vector<CellID>& remoteTargetCellsz, // Not necessary, passed for compatibility
+   vector<uint>& nPencils,
+   creal dt,
+   const uint popID,
+   Real &time
+   ) {
+
+   int trans_timer;
+   bool AMRtranslationActive = false;
+   //if (P::amrMaxSpatialRefLevel > 0) AMRtranslationActive = true;
+   // int myRank;
+   // MPI_Comm_rank(MPI_COMM_WORLD,&myRank);
+   //double t1;
+
+   trans_timer=phiprof::initializeTimer("transfer-stencil-data-all","MPI");
+   phiprof::start(trans_timer);
+   // updateRemoteVelocityBlockLists(mpiGrid,popID,VLASOV_SOLVER_NEIGHBORHOOD_ID); // already done in ACC under adjustVelocityBlocks
+   SpatialCell::set_mpi_transfer_direction(0); // Local translation uses just the X flag
+   SpatialCell::set_mpi_transfer_type(Transfer::VEL_BLOCK_DATA,false,AMRtranslationActive);
+   mpiGrid.update_copies_of_remote_neighbors(VLASOV_SOLVER_NEIGHBORHOOD_ID);
```

Comment:
Have you considered staggered communication here? Is it worth it, i.e. is the communication timer here large enough that it would benefit from "hiding" it behind a two-stage solving or is this adding one more layer of complexification that makes balancing worse?

Reply:
The issue is: we parallelize translation per-block so that we can do all local cells for a single velocity blog GlobalID at once without worrying about concurrency. If we stagger communication, we would need an inner region and an outer region, and would need to maintain two copies of the data. Not worth it.

The big imbalance and efficiency loss has been tasks sitting idle in one coordinate direction - the actual data communication is relatively fast.

Reply:
All right.